constraint-solver/src/constraint_system.rs

```diff
@@ -86,6 +91,77 @@ impl<T: RuntimeConstant, V> ConstraintSystem<T, V> {
     }
 }
 
+/// An algebraic constraint
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub struct AlgebraicConstraint<V> {
+    /// The expression representing the constraint, which must evaluate to 0 for the constraint to be satisfied.
+    pub expression: V,
+}
+
+// We implement `From` to make writing tests easier. However, we recommend using `AlgebraicConstraint::assert_zero` for clarity
+impl<V> From<V> for AlgebraicConstraint<V> {
+    fn from(expression: V) -> Self {
+        AlgebraicConstraint::assert_zero(expression)
+    }
+}
+
+impl<V> AlgebraicConstraint<V> {
+    /// Create a constraint which asserts that the expression evaluates to 0.
+    pub fn assert_zero(expression: V) -> Self {
+        AlgebraicConstraint { expression }
+    }
+}
+
+impl<V> std::ops::Deref for AlgebraicConstraint<V> {
+    type Target = V;
+
+    fn deref(&self) -> &V {
+        &self.expression
+    }
+}
+
+impl<V> std::ops::DerefMut for AlgebraicConstraint<V> {
+    fn deref_mut(&mut self) -> &mut V {
+        &mut self.expression
+    }
+}
+
+impl<T: FieldElement, V: Clone + Ord> AlgebraicConstraint<GroupedExpression<T, V>> {
+    /// Returns the referenced unknown variables. Might contain repetitions.
+    pub fn referenced_unknown_variables(&self) -> Box<dyn Iterator<Item = &V> + '_> {
+        self.expression.referenced_unknown_variables()
+    }
+
+    /// Returns a constraint which asserts that the two expressions are equal.
+    pub fn assert_eq(expression: GroupedExpression<T, V>, other: GroupedExpression<T, V>) -> Self {
```

Comment:
It could be more ergonomic to use `&GroupedExpression` here but also fine this way.